### PR TITLE
Add group update processor

### LINKED ISSUE #23546: Process `Group.UPDATE` command
<h2>Description</h2>

image

1. Create new class `GroupUpdateProcessor` and implement `DistributedTypedRecordProcessor<GroupRecord>`
2. Register the processor in the `GroupProcessors`
3. In the `processNewCommand` method:
    - Check if the group already exists. If no, reject the command with `NOT_FOUND`. If yes continue.
    - Check if the user is authorized to update groups:
        - Resource type: `GROUP`
        - Permission type: `UPDATE`
        - ResourceId: group name
    - Write a `Group.UPDATED` event and respond to the client
    - Distribute the command to the other partitions on the `DistributionQueue.IDENTITY` queueId.
4. In the `processDistributedCommand` method:
    - Write a `Group.UPDATED` event
    - Acknowledge the distribution  
5. Unit test the Group.UPDATED command
6. Unit test the Group.UPDATED command distribution

### PR DESCRIPTION
**Pull request by koevskinikola**
## Description

Add a processor for group update commands
and pass it to the appropriate engine
classes

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

closes #23546 x

### CODE REVIEW

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/group/CreateGroupMultiPartitionTest.java [hunk 1]
```diff
@@ -100,16 +99,9 @@ public void shouldDistributeInIdentityQueue() {
   public void distributionShouldNotOvertakeOtherCommandsInSameQueue() {
     // given the user creation distribution is intercepted
     for (int partitionId = 2; partitionId <= PARTITION_COUNT; partitionId++) {
-      interceptUserCreateForPartition(partitionId);
+      interceptGroupCommandForPartition(partitionId, GroupIntent.CREATE);
```
Comment (nicpuppa) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/group/CreateGroupMultiPartitionTest.java:102]: ❌ You're not testing if commands don't overtake each other here. You're only sending one command, so all that's being tested is the redistribution of that command. To test that stuff doesn't overtake each other you'll have to send another distributed command on the same queue, with a different value type or intent. Then you can guarantee it's ordered.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/group/GroupTest.java [hunk 2]
```diff
@@ -49,4 +49,38 @@ public void shouldNotDuplicate() {
...
+        engine.group().updateGroup(groupKey).withName(updatedName).update();
+
+    final var updatedGroup = updatedGroupRecord.getValue();
+    Assertions.assertThat(updatedGroup)
+        .isNotNull()
+        .hasFieldOrPropertyWithValue("name", updatedName);
+  }
+
+  @Test
+  public void shouldRejectUpdatedIfNoGroupExists() {
```
Comment (nicpuppa) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/group/GroupTest.java:72]: ❌ I'm missing the test where the processor reject the command because you are trying to update a group with the same name

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/GroupUpdateProcessor.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    if (!authCheckBehavior.isAuthorized(authorizationRequest)) {
+      final var errorMessage =
+          UNAUTHORIZED_ERROR_MESSAGE.formatted(
+              authorizationRequest.getPermissionType(), authorizationRequest.getResourceType());
+      rejectionWriter.appendRejection(command, RejectionType.UNAUTHORIZED, errorMessage);
+      responseWriter.writeRejectionOnCommand(command, RejectionType.UNAUTHORIZED, errorMessage);
+      return;
+    }
+
+    final var groupKey = groupState.getGroupKeyByName(groupName);
```
Comment (nicpuppa) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/GroupUpdateProcessor.java:80]: ❓ Is this step necessary ? Is it not enough to say that if the provided group name is the same of the persisted record the processor should reject the command ?

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/group/UpdateGroupMultiPartitionTest.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+  public void shouldDistributeInIdentityQueue() {
+    // when
+    final var name = UUID.randomUUID().toString();
+    final var groupKey = engine.group().newGroup(name).create().getKey();
+    engine.group().updateGroup(groupKey).withName(name + "-updated").update();
+
+    // then
+    assertThat(
+            RecordingExporter.commandDistributionRecords()
+                .limitByCount(r -> r.getIntent().equals(CommandDistributionIntent.FINISHED), 1)
```
Comment (nicpuppa) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/group/UpdateGroupMultiPartitionTest.java:98]: ❌ The limit should be 2

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/group/UpdateGroupMultiPartitionTest.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                .withIntent(CommandDistributionIntent.ENQUEUED))
+        .extracting(r -> r.getValue().getQueueId())
+        .containsOnly(DistributionQueue.IDENTITY.getQueueId());
+  }
+
+  @Test
+  public void distributionShouldNotOvertakeOtherCommandsInSameQueue() {
+    // given the user creation distribution is intercepted
```
Comment (nicpuppa) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/group/UpdateGroupMultiPartitionTest.java:106]: 🔧 This is not related to user creation

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/group/UpdateGroupMultiPartitionTest.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    for (int partitionId = 2; partitionId <= PARTITION_COUNT; partitionId++) {
+      interceptGroupCommandForPartition(partitionId, GroupIntent.UPDATE);
```
Comment (nicpuppa) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/group/UpdateGroupMultiPartitionTest.java:108]: ❌ You should intercept the Group creation so you can verify that the distribution do not overtake other commands in the queue

**Review by nicpuppa — CHANGES_REQUESTED** [review]
Hey @koevskinikola I left some comments, please have a look

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/GroupUpdateProcessor.java (hunk 3 above)
Comment (koevskinikola) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/GroupUpdateProcessor.java:80]: That's a very good point, thanks!

I'll update this processor and the`RoleUpdateProcessor` to avoid this extra state access.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/group/CreateGroupMultiPartitionTest.java (hunk 1 above)
Comment (koevskinikola) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/group/CreateGroupMultiPartitionTest.java:102]: @nicpuppa I'm creating another group on line 108, so two `GroupIntent.CREATE` commands are sent.

Or did I misunderstand something?

**Comment by koevskinikola** [comment]
@nicpuppa I implemented your feedback except [this one](https://github.com/camunda/camunda/pull/24342#discussion_r1829003673), let me know how it looks now.

**Review by nicpuppa — APPROVED** [review]
Thanks @koevskinikola

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/GroupUpdateProcessor.java (hunk 3 above)
Comment (remcowesterhoud) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/GroupUpdateProcessor.java]: @nicpuppa @koevskinikola This was necessary. If the group name changed we must check the state that there is not already another group with this name. Checking it with the provided record is not the same thing.

Comment (nicpuppa) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/GroupUpdateProcessor.java]: What do you mean ? We check the provided with the record in the state

Comment (remcowesterhoud) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/GroupUpdateProcessor.java]: > What do you mean ? We check the provided with the record in the state

We get the Group from the state using the key provided in the command. Then we check if the name in our state is the same as the name of the provided command.

What the check should do is make sure there are no other Groups (with different keys) that have the same name. The name should be unique.

Comment (remcowesterhoud) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/GroupUpdateProcessor.java]: So basically:

```java
if (groupState.getByName(name).isPresent()) {
   // reject because a group with this name already exists.
}
```

Comment (nicpuppa) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/GroupUpdateProcessor.java]: Ok got it, make sense 👍

Comment (remcowesterhoud) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/GroupUpdateProcessor.java]: @nicpuppa Will you create a fix for it?

Comment (nicpuppa) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/GroupUpdateProcessor.java]: yes
